src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

```diff
@@ -1531,4 +1533,113 @@ private OpenIdConnectProtocolException CreateOpenIdConnectProtocolException(Open
         ex.Data["error_uri"] = errorUri;
         return ex;
     }
+
+    /// <summary>
+    /// Handles the `ValidatePrincipal event fired from the underlying CookieAuthenticationHandler.
+    /// This is used for refreshing OIDC auth. token if needed.
+    /// </summary>
+    /// <param name="context">The CookieValidatePrincipalContext passed as part of the event.</param>
+    internal static async Task ValidatePrincipal(CookieValidatePrincipalContext context)
+    {
+        var authHandlerProvider = context.HttpContext.RequestServices.GetRequiredService<IAuthenticationHandlerProvider>();
+        var handler = await authHandlerProvider.GetHandlerAsync(context.HttpContext, context.Scheme.Name);
+        if (handler is OpenIdConnectHandler oidcHandler)
+        {
+            await oidcHandler.HandleValidatePrincipalAsync(context);
+        }
+    }
+
+    private async Task HandleValidatePrincipalAsync(CookieValidatePrincipalContext validateContext)
+    {
+        var accessTokenExpirationText = validateContext.Properties.GetTokenValue("expires_at");
+        if (!DateTimeOffset.TryParse(accessTokenExpirationText, out var accessTokenExpiration))
+        {
+            return;
+        }
+
+        var oidcOptions = this.OptionsMonitor.Get(validateContext.Scheme.Name);
+        var now = oidcOptions.TimeProvider!.GetUtcNow();
+        if (now + TimeSpan.FromMinutes(5) < accessTokenExpiration)
+        {
+            return;
+        }
+
+        var tokenRefreshContext = new Events.TokenRefreshContext(Context, Scheme, oidcOptions, validateContext.Principal!, validateContext.Properties);
+        await Options.Events.TokenRefreshing(tokenRefreshContext);
+        if (tokenRefreshContext.ShouldRefresh)
```

Comment:
Should the `tokenRefreshContext.Result` be checked like the most of the events? Something like this: 

```
if (tokenRefreshContext.Result != null)
{
    if (tokenRefreshContext.Result.Handled)
    {
        Logger.TokenRefreshingHandledResponse();
    }
    else if (tokenRefreshContext.Result.Skipped)
    {
        Logger.TokenRefreshingSkipped();
    }
}
```